[jscodeshift] Add types for jscodeshift 0.6

brieb opened the pull request:
[Jscodeshift](https://github.com/facebook/jscodeshift/) is a very popular tool for writing codemods (ast-based transforms of code). 

Here is an example of a jscodeshift transform (taken from the default contents of the transform panel on [AST Explorer](https://astexplorer.net/)). This code reverses the names of all identifiers (indeed useless, but certainly fun 🙂).

```
function transform(file, { j }) {
  return j(file.source)
    .find(j.Identifier)
    .forEach(path => {
      j(path).replaceWith(
        j.identifier(path.node.name.split('').reverse().join(''))
      );
    })
    .toSource();
}
```

Even in this simple example, we can start to see why TypeScript would be beneficial.
`j.Identifier` is a kind of node and there are [hundreds of node kinds](https://github.com/benjamn/ast-types/blob/master/gen/nodes.ts).
`j.identifier(...)` is a node builder and there are [hundreds of those too](https://github.com/benjamn/ast-types/blob/master/gen/builders.ts) with many function arguments.
The autocomplete and type-checking benefits for nodes and builders, as well as for the jscodeshift api in general has made me much more productive when writing codemods (with a local copy of this d.ts file). And had this been around when I got started with jscodeshift, it would have made it more approachable and easier to learn.

Here is a demonstration of TypeScript in action with the above example.

![jscodeshift-ts](https://user-images.githubusercontent.com/723547/48811967-02984680-ece5-11e8-8821-4c33c0646288.gif)

Jscodeshift depends on recast and ast-types and they make up a significant part of its API. I worked with @benjamn to convert the [recast](https://github.com/benjamn/recast/pull/554) and [ast-types](https://github.com/benjamn/ast-types/pull/300) repos to TypeScript. This PR uses those type annotations and adds types for jscodeshift. This PR the final step to a TypeScript-powered developer experience for writing jscodeshift codemods.

---

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

cc @benjamn @pspeter3 @fkling @cpojer 

---

### Tests:

- Basic cases: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/32020/files#diff-6761e67434cfd48e6684d3db09029787
- Sample jscodeshift transforms taken from https://github.com/jhgg/js-transforms and converted to TypeScript, for use as real-world test cases: https://github.com/DefinitelyTyped/DefinitelyTyped/tree/d542e7c4eaa682a01e4ce430a366444fba334478/types/jscodeshift/test/js-transforms

pspeter3 commented:
This looks awesome! I'm really excited for this to get merged. It seems like we also need to add `ast-types` to `dependenciesWhitelist.txt` in order for this to get merged?

brieb commented:
Have a PR open for that here https://github.com/Microsoft/types-publisher/pull/550

brieb commented:
@typescript-bot Hoping that commenting will keep this PR alive. Build failure is blocked on https://github.com/Microsoft/types-publisher/pull/550

brieb commented:
@PranavSenthilnathan I updated these definitions, as per your comments:
- split into multiple files
- use `export =` to be consistent with the JS implementation
Thank you for taking a look!

brieb commented on types/jscodeshift/src/core.d.ts:
```diff
@@ -0,0 +1,363 @@
+
+    type ASTNode = nodes.ASTNode;
+
+    type AnyTypeAnnotation = nodes.AnyTypeAnnotation;
```
Before, I had used `export * from "ast-types/gen/nodes";` for this, but I couldn't get that working after changing this module to use `export =`. Is there a better way to do this?

brieb commented:
Oh shoot - looks like a newer version of ast-types (0.12.2) was released that broke CI. Things were working locally with 0.12.1 installed. Sorry. I'll fix these errors...

PranavSenthilnathan commented:
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!
